### ISSUE TITLE
Exception in training due to unknown words

### ISSUE THREAD

**Issue by marnaoty** [danhper/suplearn-clone-detection#5]
I was trying to replicate your work (using your and my datasets) using the directions you provided through your GitHub account.
However, when I start training I get exceptions like:

2019-07-03 18:20:44,702 ERROR failed: ('NameStore', None)
...
2019-07-04 13:01:53,746 ERROR failed: ('ImportDeclaration', None)

I used the json files on your clone-detection repository to generate the .tsv vocabulary files for both java and python with different vocabulary sizes. But I always get exceptions similar to the above.

I appreciate your help, and I can post my config files or anything you might ask about.

**Comment by marnaoty** [comment]
Daniel Reply:

Looking at the error, I would assume that a word is missing from the vocabulary. I suppose what you are seeing is the text from the `KeyError` exception. You can add `--debug` before the sub command so it will show the stacktrace if needed.
Now, I'm not entirely sure why a word is missing. What comes to mind would be you are feeding in the wrong language. If the model is expecting Java-Python pairs, all the pairs must be Java-Python, and it will fail if you try to pass Python-Java or Java-Java.
Different vocabulary size should not be an issue.

**Comment by marnaoty** [comment]
"If the model is expecting Java-Python pairs, all the pairs must be Java-Python, and it will fail if you try to pass Python-Java or Java-Java."

How do you decide that order between java and python? through the configuration files?

**Comment by danhper** [comment]
> How do you decide that order between java and python? through the configuration files?

Yes, the `languages` key of the config file is used for this.
https://github.com/danhper/suplearn-clone-detection/blob/master/config.yml.example#L26
`languages` must be a two elements array, each containing config for a language. In this example, the first language is `java` and the other is `python`, so the pairs must be Java-Python.

**Comment by marnaoty** [comment]
Then I believe it is in the correct order.
I also generate samples using your tool "suplearn-clone generate-dataset -c config.yml"

Here is my configuration file content
[config.txt](https://github.com/danhper/suplearn-clone-detection/files/3373826/config.txt)

**Comment by danhper** [comment]
Ok, I think I understood what is wrong. The issue is mainly with the lack of documentation, sorry about that.
When a pair `(node_type, node_value)` is not present in the vocabulary, the model tries to look for `(node_type, None)` and fails if not found.
By default, the vocabulary generator of `bigcode-ast-tools` does not systematically include  `(node_type, None)` pairs. The `--include-types` option can be passed to the `generate-vocabulary` command of `bigcode-ast-tools` to force it to include these pairs.
Could you try to re-generate the vocabulary using the `--include-types` option please?

**Comment by marnaoty** [comment]
Unfortunately, this didn't solve the problem completely.

I still get some of these (node_type,None) pairs but with different values than before. e.g. ('CompareGtEGtE', None)

When I trained with no-ids in the vocabulary I got the same exception but with little different output:

2019-07-09 15:42:24,515 ERROR failed: ('CompareEqEqEq',)

**Comment by danhper** [comment]
I see. This type of node should have been normalized when generating the AST (https://github.com/danhper/bigcode-tools/tree/master/bigcode-astgen/python) I'm not sure why it was not in the current DB.
I regenerated the DB with the normalized AST, could you give it a try please?
You should be able to download it from here:
https://daniel.perez.sh/research/2019/cross-language-clones/

**Comment by marnaoty** [comment]
I downloaded the DB again, I didn't generate new samples since the table was already populated.
I tried training again and got:

Epoch 1/10
2019-07-09 19:27:19,887 ERROR failed: 'ProgbarLogger' object has no attribute 'log_values'

**Comment by danhper** [comment]
The error message deserves some improvement but you do need to run `suplearn-clone generate-dataset -c config.yml` first.

**Comment by marnaoty** [comment]
I regenerated and still get the same exceptions:

2019-07-10 11:43:44,904 ERROR failed: ('BinOpFloorDiv', None)

**Comment by marnaoty** [comment]
Wait, I think I'm generating using the wrong json files.
I will try and get back to you

**Comment by marnaoty** [comment]
I guess the main issue is the mismatch between the json files and the generated dataset.
I have already used [bigcode](https://github.com/danhper/bigcode-tools/blob/master/doc/tutorial.md) tool to generate the ast files for both java and python.
But I didn't change the config file to refer to these new json files.
Now I'm a little bit confused.
The config file has only one ast_path under the generator. Does that mean I should concatenate both java asts and python asts in one file?
Also, the submission file. I think it is the same as info.json file that I got from the bigcode tool. But I got that file for python but not for java.
Note: the command for python was not documented, so I had to guess it from the source files, and it worked.

**Comment by danhper** [comment]
> I regenerated and still get the same exceptions:
> 2019-07-10 11:43:44,904 ERROR failed: ('BinOpFloorDiv', None)

How did you generate the vocabulary? `BinOpFloorDiv` should definitely exist. For example,

```
echo '1 // 2' > sample.py
bigcode-astgen-py sample.py > sample-ast.json
bigcode-ast-tools generate-vocabulary sample-ast.json -o sample-vocab.tsv --include-types
```

should yield the following:

```tsv
id      type    metaType        count   value
0       Num     Other   2
1       Expr    Expr    2
2       Module  Other   2
3       BinOpFloorDiv   Other   2
4       Num     Other   1       "2"
5       Num     Other   1       "1"
```

> I have already used bigcode tool to generate the ast files for both java and python.

Sorry about the confusion.
The JSON files for the clone dataset should not be needed, everything is already in the SQLite database. If you do want to re-generate the database from scratch, then you will need to concat the Python and Java AST files, so something like

```
cat java-asts.txt python-asts.txt > asts.txt
cat java-asts.json python-asts.json > asts.json
```

and the `scripts/create_submissions.py` can then be used to create the SQLite database (containing only the submissions, not the sample pairs).
Again, this step should not be needed, as the data should already be in the SQLite database.

> Note: the command for python was not documented, so I had to guess it from the source files, and it worked.

Which command are you talking about? `bigcode-asgten-py`? If so, the documentation is here https://github.com/danhper/bigcode-tools/tree/master/bigcode-astgen/python and the `--help` should also provide some information. If you feel that something is missing from the documentation, a PR would be more than welcome.

**Comment by marnaoty** [comment]
Sorry for my late reply.

"The JSON files for the clone dataset should not be needed, everything is already in the SQLite database."

Ok, So, JSON files information is already in the DB, I don't need to regenerate them using bigcode tool. Cool.

Yet, we still need the vocabulary ".tsv" files for training data for both java and python.
I assume that I can obtain these files only by running bigcode-ast-tools generate-vocabulary on the ast JSON files.
So, even though the contents of the JSON files are included in the DB, I still need the JSON files to generate the vocabulary files. Right?

**Comment by marnaoty** [comment]
BTW, BinOpFloorDiv does exist in the python vocabulary and not the java vocabulary.
I might have generated the vocabulary files incorrectly by using the wrong json files as input.
I think the key to solving this issue is vocabulary generation.

**Comment by danhper** [comment]
> So, even though the contents of the JSON files are included in the DB, I still need the JSON files to generate the vocabulary files. Right?

I usually use a different set of files to generate the vocabulary as I also pre-train embeddings, and this requires more data but you can of course use the AST JSON files.

Did you add the `--include-types` option when creating the vocabulary?

**Comment by marnaoty** [comment]
yes, I used --include-types.
The json files I have generated first from the java and python datasets [here](https://www.csg.ci.i.u-tokyo.ac.jp/projects/clone/).
Then I generated vocab using the following commands:

docker-bigcode bigcode-ast-tools generate-vocabulary --size 10000 --include-types -o workspace/py-vocabulary-10000.tsv workspace/python-projects-asts.json  
docker-bigcode bigcode-ast-tools generate-vocabulary --size 10000 --include-types -o workspace/java-vocabulary-10000.tsv workspace/apache-commons-asts.json

If you have the vocabulary files already generated, or the pre-train embedding files, that would be a great help.

**Comment by danhper** [comment]
I generated the vocabulary using this dataset I and uploaded it to a gist.
https://gist.github.com/danhper/2d8257cc954a8ba530631186edd35d28

The steps are mostly what is documented and I did not get any error when trying to train:

```
bigcode-astgen-py --batch 'repositories/**/*.py' -o python-asts
bigcode-astgen-java --batch 'repositories/**/*.java' -o java-asts
bigcode-ast-tools generate-vocabulary -s 10000 --include-types python-asts.json -o python-vocab.tsv
bigcode-ast-tools generate-vocabulary -s 10000 --include-types java-asts.json -o java-vocab.tsv
```

**Comment by marnaoty** [comment]
yes, the training finally worked after I used your tsv files and regenerated the samples.
Thank you for your help.
I think the major problem was the mismatch between the database and the tsv files. In other words, the source from which the tsv files were generated in my case wasn't the same from which the DB is created.
Anyway thank you so much.
